lib/FIR/statements.h

```diff
@@ -556,21 +580,12 @@ class Statement : public SumTypeMixin<ReturnStmt,  //
     parent->insertBefore(this);
   }
   std::string dump() const;
-
-  // g++/clang++ will optimize this to a simple register copy
-  // Every Stmt_impl is wrapped in and the first data member of a Statement;
-  // therefore, a pointer to one or the other is bitwise identical.
-  // This checks that this assumption is, in fact, true.
-  static Statement *From(Stmt_impl *stmt) {
-    static Statement s{nullptr, UnreachableStmt::Create()};
-    auto *result{reinterpret_cast<Statement *>(reinterpret_cast<char *>(stmt) -
-        (reinterpret_cast<char *>(&s.u) - reinterpret_cast<char *>(&s)))};
-    CHECK(result == reinterpret_cast<Statement *>(stmt) &&
-        "expecting pointers to be equal");
-    return result;
-  }
 };
 
+template<typename A> inline QualifiedStmt<A>::operator A *() const {
+  return reinterpret_cast<A *>(&stmt->u);
```

Comment:
Because it's actually not what I want.  `get` will work if we want an exact type, but not if we want any of a set of subtypes.

Reply:
The second option should also work with subtypes (if these are things like `TerminatorStmt_impl`, `ActionStmt_impl`...). But I agree it may be a bit bad looking and does more stuff than a `reinterpret_cast`, though these things are mostly checks.

Reply:
OK, It still seems like over-engineering to add run-time tests to confirm that the compile-time constraints actually do what they are designed for.

Reply:
Casting a pointer to a std::variant<> into a pointer to a distinct std::variant<> type assumes that the two variant types use the same tag codes for their shared alternatives.